File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <functional>
#include <queue>
#include <unordered_map>
#include <algorithm>
#include <iostream>

using namespace std;

vector<string> solution(vector<vector<string>> tickets) {
    vector<string> answer = {"ICN"};
    unordered_map<string,vector<pair<string, bool>>> m;
    for(auto t: tickets) {
        string a = t[0];
        string b = t[1];
        m[a].push_back({b, false});
    } // icn : sfo,false
    
    for(auto& [key, val]: m) {
        sort(val.begin(), val.end(), [](auto a, auto b) {
            return a.first < b.first;
        });
    }
    
    bool found = false;
    
    function<void(string)> dfs = [&](string curr) {
        if (found) return;
        if (answer.size() == tickets.size() + 1) {
            found = true;
            return;
        }
        for(auto& next: m[curr]) {
            if(!next.second) {
                next.second = true;
                answer.push_back(next.first);

                dfs(next.first);
                if (found) return;

                next.second = false;
                answer.pop_back();
            }
        }
    };
    
    
    dfs("ICN");
    
    return answer;
}

File: 프로그래머스/3/43164. 여행경로/여행경로.swift
import Foundation

func solution(_ tickets:[[String]]) -> [String] {
    let tickets = tickets.sorted {
        if $0[0] == $1[0] {
            return $0[1] < $1[1]
        }
        return $0[0] < $1[0]
    }
    
    var dict: [String:[String]] = [:]
    for t in tickets {
        dict[t[0], default:[]].append(t[1])
    }
    
    var visited = Array(repeating: false, count: tickets.count)
    var answer = [String]()
    
    func dfs(_ st: String, _ route: [String]) -> Bool {
        if route.count == tickets.count + 1 {
            answer = route
            return true
        }
        
        for i in 0..<tickets.count {
            if !visited[i] && tickets[i][0] == st {
                visited[i] = true
                if dfs(tickets[i][1], route + [tickets[i][1]]) {
                    return true
                }
                visited[i] = false
            }
        }
        return false
    }
    _ = dfs("ICN", ["ICN"])
    return answer
}